lsp/src/Language/LSP/Server/Processing.hs

```diff
@@ -449,31 +451,31 @@ handle' ::
   TClientMessage meth ->
   m ()
 handle' logger mAction m msg = do
-  maybe (return ()) (\f -> f msg) mAction
+  shutdown <- isShuttingDown
+  let allowedMethod m = case (splitClientMethod m, m) of
+        (IsClientNot, SMethod_Exit) -> True
+        (IsClientReq, SMethod_Shutdown) -> True
+        _ -> False
+
+  logger <& LspMessage (show m) `WithSeverity` Debug
+  when (not shutdown || allowedMethod m) $ maybe (return ()) (\f -> f msg) mAction
```

Comment:
I think this would be stronger than it needs to be. In particular, it's not a problem if we e.g. update the VFS. It's really only the config callback that's potentially problematic.

It's also not enough, since updating the configuration is complicated. This handles the `didChangeConfiguration` notification, but not the response to `workspace/configuration`. So if we wanted to do this, I think we should just stick the check in `tryUpdateConfig` or something.

And I still think that servers should just make their config callbacks safe :)

Reply:
You are right , it is not enough to stop the callback.
But I do not see any reason to do the `action before noti/req` even it is safe to do so, since the server is marked as shutdown anyway(Waitting the cleanup and exit notification).🤔

Reply:
Is this correct ?

There is push model and pull model(new one) in LSP,
The pull model:
1. `didChangeConfiguration ` sent to server 
2. `workspace/configuration` sent to client
3.  Response of configuration sent to server, we update the config.

What we are doing here is kind of mixed:
1. `didChangeConfiguration ` sent to server, we `tryChangeConfig`.
2. `workspace/configuration` sent to client
3.  Response of configuration sent to server, `tryChangeConfig` again.